appcelerator-archive/amp
# Make function workers deploy globally on worker nodes

## How to test
```
ampmake clean build
alias amp='amp -s localhost'
amp -cluster create --tag=local
. ./hack/ampenv
amp fn create banner bquenin/banner-function
echo 'ATOMIQ rocks!' | curl -k https://faas.local.atomiq.io/banner --data-binary @-
echo 'ATOMIQ rocks!' | curl -k https://faas.local.atomiq.io/banner --data-binary @-
echo 'ATOMIQ rocks!' | curl -k https://faas.local.atomiq.io/banner --data-binary @-
amps service logs amp_funcexec
```

Make sure from the logs that the functions are executed alternatively on workers, such as this:
```
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Container created: ddeb1f7c81e82421e97091b4a78396506e85451bbb027549c4397f54d18f2359
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Container started
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Container attached
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Transferred standard input (14 bytes)
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Transferred standard output (378 bytes)
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Function call executed
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:38 Function return successfuly submitted, call Id: 5abd0565c1f36f128ba1f7cc8087bd0ff7520394c5af8e692bf65af25f73aa2b
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:39 Function call received, call id: 31ec095a5a4efb5b7cf520adfa30db2430c76c73d1d5a90a7689fc48a82c1c5a
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:40 Container created: f8b7f5c15541ef3c98a6326bff4ca015c24cdec0c697b371033ffd4fb8f53a3e
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:40 Container started
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:40 Container attached
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:40 Transferred standard input (14 bytes)
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:41 Transferred standard output (378 bytes)
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:41 Function call executed
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:41 Function return successfuly submitted, call Id: 31ec095a5a4efb5b7cf520adfa30db2430c76c73d1d5a90a7689fc48a82c1c5a
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:42 Function call received, call id: 98438cf07e965d2c62130b90dd1139196eaf24fc3f8d2f7cdf7b61806a65dc5e
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:42 Container created: f8a6c86fe22c5d607221bd5df118194fa5f5c48da3278f1ae19bafb77bbea6f7
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:42 Container started
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:42 Container attached
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:42 Transferred standard input (14 bytes)
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:43 Transferred standard output (378 bytes)
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:43 Function call executed
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:43 Function return successfuly submitted, call Id: 98438cf07e965d2c62130b90dd1139196eaf24fc3f8d2f7cdf7b61806a65dc5e
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:44 Function call received, call id: 16a57ad90b2d7616f9a2f391fb3444553978dc4d2e2cd76b49f96d0fe4738207
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:44 Container created: 46b83daec771f8559b27a936bdc04a309b9996a1d4ee2702200d4a05397ccb69
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:44 Container started
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:44 Container attached
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:44 Transferred standard input (14 bytes)
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:45 Transferred standard output (378 bytes)
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:45 Function call executed
amp_funcexec.0.2x9vqgi1smbb@75717d43aff3    | 2017/04/25 23:50:45 Function return successfuly submitted, call Id: 16a57ad90b2d7616f9a2f391fb3444553978dc4d2e2cd76b49f96d0fe4738207
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:46 Function call received, call id: 66d748fb39cb3d2b7fd6a7e840b4467f1a6c3f2ded31ac3fcab28e14e38b1523
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:46 Container created: 2da187c35940482192e17f7605e7ea6734eee49560f85227f1e1b7f04a85a40c
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:47 Container started
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:47 Container attached
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:47 Transferred standard input (14 bytes)
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:47 Transferred standard output (378 bytes)
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:47 Function call executed
amp_funcexec.0.ukojzpf8usdw@a92191f714b0    | 2017/04/25 23:50:47 Function return successfuly submitted, call Id: 66d748fb39cb3d2b7fd6a7e840b4467f1a6c3f2ded31ac3fcab28e14e38b1523
```
